blockchain/event_filter.go

```diff
@@ -0,0 +1,198 @@
+	To
+)
+
+func newEventFilter(txn db.Transaction, from *felt.Felt, keys []*felt.Felt, fromBlock, toBlock uint64) *EventFilter {
+	return &EventFilter{
+		txn:       txn,
+		From:      from,
+		Keys:      keys,
+		FromBlock: fromBlock,
+		ToBlock:   toBlock,
+	}
+}
+
+// SetRangeEndBlockByNumber sets an end of the block range by block number
+func (f *EventFilter) SetRangeEndBlockByNumber(end FilterRangeEnd, blockNumber uint64) error {
+	_, err := blockHeaderByNumber(f.txn, blockNumber)
+	if err != nil {
+		return err
+	}
+	if end == From {
+		f.FromBlock = blockNumber
+	} else if end == To {
+		f.ToBlock = blockNumber
+	} else {
+		return errors.New("undefined range end")
+	}
+	return nil
+}
+
+// SetRangeEndBlockByHash sets an end of the block range by block hash
+func (f *EventFilter) SetRangeEndBlockByHash(end FilterRangeEnd, blockHash *felt.Felt) error {
+	header, err := blockHeaderByHash(f.txn, blockHash)
+	if err != nil {
+		return err
+	}
+	return f.SetRangeEndBlockByNumber(end, header.Number)
+}
+
+// Close closes the underlying database transaction that provides the blockchain snapshot
+func (f *EventFilter) Close() error {
+	return f.txn.Discard()
+}
+
+type ContinuationToken struct {
+	fromBlock       uint64
+	processedEvents uint64
+}
+
+func (t *ContinuationToken) String() string {
+	return fmt.Sprintf("%d-%d", t.fromBlock, t.processedEvents)
+}
+
+func (t *ContinuationToken) FromString(str string) error {
+	_, err := fmt.Sscanf(str, "%d-%d", &t.fromBlock, &t.processedEvents)
+	return err
+}
+
+type FilteredEvent struct {
+	*core.Event
+	BlockNumber     uint64
+	BlockHash       *felt.Felt
+	TransactionHash *felt.Felt
+}
+
+func (f *EventFilter) Events(cToken *ContinuationToken, chunkSize uint64) ([]*FilteredEvent, *ContinuationToken, error) {
+	var matchedEvents []*FilteredEvent
+
+	filterKeysMap := make(map[felt.Felt]bool, len(f.Keys))
+	for _, key := range f.Keys {
+		filterKeysMap[*key] = true
+	}
+
+	for curBlock := f.FromBlock; curBlock <= f.ToBlock; curBlock++ {
+		// skip the blocks that we previously processed for this request
+		if cToken != nil && curBlock < cToken.fromBlock {
+			continue
+		}
+
+		header, err := blockHeaderByNumber(f.txn, curBlock)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// test `from` only by default if it's not nil because empty filter keys means match all
+		possiblyMatches := true
+		if f.From != nil {
+			fromBytes := f.From.Bytes()
+			possiblyMatches = header.EventsBloom.Test(fromBytes[:])
+			// bloom filter says no events from this contract
+			if !possiblyMatches {
+				continue
+			}
+		}
+		for key := range filterKeysMap {
+			keyBytes := key.Bytes()
+
+			// check if block possibly contains the event we are looking for
+			possiblyMatches = header.EventsBloom.Test(keyBytes[:])
+			if possiblyMatches {
+				break
+			}
+		}
+
+		// bloom filter says no events match the filter, skip this block entirely if from is not nil
+		if !possiblyMatches {
+			continue
+		}
```

Comment:
It took me a second to connect this to `core.EventsBloom`. I feel the two snippets should be close to each other in the codebase, since they change together.

What do you think about an `events` package?

- Exports: `Filterer` interface, `Filter` struct, and `Query` struct
- Create a filter with: `events.NewFilter(r blockchain.Reader) *Filter`
- Execute a query with: `filter.Filter(q *Query) []*core.Event`
- Calculate bloom filter with: `events.Bloom(receipts []*core.TransactionReceipt) *bloom.BloomFilter` 

The `rpc` package can manage the continuation token details, while the blockchain is behind an interface and decoupled from event filtering.

Reply:
Why I had to go with this implementation was consistency. All DB access should be handled by a single db.Transaction or we risk returning inconsistent results. If we rely on blockchain.Reader interface to access the DB, it will create separate Transactions for each request. That is why EventFilter has a Transaction reference and uses private blockchain package functions that operate within a Transaction context.

By letting Blockchain create the EventFilter we are ensuring the entire operation is atomicly executed within a single db.Transaction context and the result is consistent.

Think of the scenario where user defines the range in terms of hashes. Between we verify that block hash exists and actually run the filter on that block, that block could be reorged and leave us in a weird situation.

or think of the scenario where we are mid way filtering the events and the entire range that user defined gets reorged. Half of our results will be pre-reorg, the other half will be post-reorg.

Reply:
I think completely abstracting DB behind the Blockchain still makes sense. But regardless, that would be a huge refactor, maybe in a separate PR after giving it some consideration.